File: InterviewCakeProblems2/mergeMeetingTimes2.js
function mergeMeetingTimes(meetingTimes){
    sortedMeetingTimes = meetingTimes.sort( (a,b) => a.startTime - b.startTime);
    
    mergedMeetingTimes = [sortedMeetingTimes[0]];

    for (let i = 1; i < sortedMeetingTimes.length; i++) {
        let currentMeeting = sortedMeetingTimes[i];
        let lastMergedMeetingTime = mergedMeetingTimes[mergedMeetingTimes.length - 1]
        if (currentMeeting.startTime <= lastMergedMeetingTime.endTime) {
            lastMergedMeetingTime.endTime = Math.max(currentMeeting.endTime, lastMergedMeetingTime.endTime);
        } else {
            mergedMeetingTimes.push(currentMeeting)
        }
    }
    return mergedMeetingTimes
}

meetingTimes1 = [{ startTime: 2, endTime: 3 }, { startTime: 6, endTime: 9 }]
meetingTimes2 = [
    { startTime: 0, endTime: 1 },
    { startTime: 3, endTime: 5 },
    { startTime: 4, endTime: 8 },
    { startTime: 10, endTime: 12 },
    { startTime: 9, endTime: 10 },
]
meetingTimes3 = [{ startTime: 1, endTime: 2 }, { startTime: 2, endTime: 3 }];
meetingTimes4 = [{ startTime: 1, endTime: 5 }, { startTime: 2, endTime: 3 }];
meetingTimes5 = [
    { startTime: 1, endTime: 10 },
    { startTime: 2, endTime: 6 },
    { startTime: 3, endTime: 5 },
    { startTime: 7, endTime: 9 },
];

console.log(mergeMeetingTimes(meetingTimes1));
console.log(mergeMeetingTimes(meetingTimes2));
console.log(mergeMeetingTimes(meetingTimes3));
console.log(mergeMeetingTimes(meetingTimes4));
console.log(mergeMeetingTimes(meetingTimes5));

File: InterviewCakeProblems2/merge_meeting_times2.py
def merge_meeting_times(meeting_times):
    sorted_meeting_times = sorted(meeting_times)
    merged_meetings = [sorted_meeting_times[0]]

    for current_meeting_start, current_meeting_end in sorted_meeting_times[1:]:
        last_merged_meeting_start, last_merged_meeting_end = merged_meetings[-1]
        if current_meeting_start <= last_merged_meeting_end:
            merged_meetings[-1] = (last_merged_meeting_start, max(last_merged_meeting_end, current_meeting_end))
        else:
            merged_meetings.append((current_meeting_start, current_meeting_end))
    return merged_meetings

meeting_times1 = [(2, 3), (6, 9)]
meeting_times2 = [
    (0, 1),
    (3, 5),
    (4, 8),
    (10, 12),
    (9, 10)
]
meeting_times3 = [(1, 2), (2, 3)]
meeting_times4 = [(1, 5), (2, 3)]
meeting_times5 = [
    (1, 10),
    (2, 6),
    (3, 5),
    (7, 9)
]

print(merge_meeting_times(meeting_times1))
print(merge_meeting_times(meeting_times2))
print(merge_meeting_times(meeting_times3))
print(merge_meeting_times(meeting_times4))
print(merge_meeting_times(meeting_times5))
